### PR TITLE
CHNS example : capillary rise and Jurin's law

### PR DESCRIPTION
**Pull request by PierreLaurentinCS**
### Description

This is the first example of the Cahn-Hilliard-Navier-Stokes solver. It displays the use of angle of contact boundary conditions to solve a capillary rise problem in 2D. The numerical height difference is compared to that predicted by the Jurin's law and show excellent agreement . The pressure fields are qualitatively commented and show that the solver adequately describe pressure jumps due to an interface.

A post processing python script is included with the example to reproduce the figures if needed. They were added in a previous PR (https://github.com/chaos-polymtl/lethe/pull/1307) and already reviewed.

**EDIT**
There was a modification made to the implementation of the boundary conditions to make the example work with the new boundary conditions.
An application test (cahn_hilliard_navier_stokes_jurins_law_2d) was added to ensure of the future stability of the solver

### Checklist (will be removed when merged)

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Copyright headers are present and up to date
- [x] Lethe documentation is up to date
- [x] The branch is rebased onto master
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] Changelog is modified
- [x] The PR description is cleaned and ready for merge

### CODE REVIEW

File: doc/source/examples/multiphysics/jurins-law/jurins-law.rst [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+        set method    = automatic
+        set verbosity = verbose
+      end
+    end
+
+
+Initial Conditions
+~~~~~~~~~~~~~~~~~~
+
+In the ``initial conditions`` subsection, we need only need to initialize the phase field in the ``cahn hilliard`` subsection. The chemical potential field is set to :math:`0` uniformly. The interface is initialized with the equilibrium interface thickness, which requires to know the value of :math:`\epsilon` that is outputed at every iteration.
```
Comment (AmishgaAlphonius) [inline doc/source/examples/multiphysics/jurins-law/jurins-law.rst:185]:
```suggestion
In the ``initial conditions`` subsection, we need only need to initialize the phase field in the ``cahn hilliard`` subsection. The chemical potential field is set to :math:`0` uniformly. The interface is initialized with the equilibrium interface thickness, which requires to know the value of :math:`\epsilon` that is outputted at every iteration. Here, :math:`\epsilon = 0.04419`.
```

File: doc/source/examples/multiphysics/jurins-law/jurins-law.rst [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+|                                                                                                                   |
+|     Pressure fields at the end of the simulation for different angles of contact.                                 |
+|                                                                                                                   |
++-------------------------------------------------------------------------------------------------------------------+
+
+---------------------------
+Possibilities for Extension
+---------------------------
+
+- **Going 3D**: the mesh can be extruded into the third dimension and there is an adaptation of the Jurin's law in three dimensions. Some results are available in the literature for comparison (see Lovrić et al. [#lovric2019]_)
```
Comment (AmishgaAlphonius) [inline doc/source/examples/multiphysics/jurins-law/jurins-law.rst:296]:
```suggestion
- **Going 3D**: the mesh can be extruded into the third dimension and there is an adaptation of the Jurin's law in three dimensions. Some results are available in the literature for comparison (see Lovrić *et al.* [#lovric2019]_)
```

File: doc/source/examples/multiphysics/jurins-law/jurins-law.rst [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+- **Investigate the effect of a no-slip boundary condition**: instead of the slip boundary condition imposed on the inner face of the wall, we could try to use a no-slip boundary condition. This situation would be closer to a real capillary rise experiment. We expect to observe a different transitory state with this new boundary condition.
```
Comment (AmishgaAlphonius) [inline doc/source/examples/multiphysics/jurins-law/jurins-law.rst:298]:
```suggestion
- **Investigate the effect of a no-slip boundary condition**: instead of the slip boundary condition imposed on the inner face of the wall, we could try to use a no-slip boundary condition. This situation would be closer to a real capillary rise experiment. We expect to observe a different transient state with this new boundary condition.
```

File: doc/source/examples/multiphysics/jurins-law/images/final_pressure_gradient.png
Comment (AmishgaAlphonius) [inline doc/source/examples/multiphysics/jurins-law/images/final_pressure_gradient.png:1]: file not used

File: doc/source/examples/multiphysics/jurins-law/images/pressure_difference.svg
Comment (AmishgaAlphonius) [inline doc/source/examples/multiphysics/jurins-law/images/pressure_difference.svg:1]: file not used

File: doc/source/examples/multiphysics/jurins-law/images/pressure_difference.png
Comment (AmishgaAlphonius) [inline doc/source/examples/multiphysics/jurins-law/images/pressure_difference.png:1]: If you have the state file not too far, you could add the units here.

File: examples/multiphysics/jurins-law-2d/jurins-law-2d.prm
Comment (AmishgaAlphonius) [inline examples/multiphysics/jurins-law-2d/jurins-law-2d.prm:1]: line 20: end time should be 0.5

**Review by AmishgaAlphonius — COMMENTED** [review]
Great work! :)

File: doc/source/examples/multiphysics/jurins-law/jurins-law.rst [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+|     Representation of the initial (left) and final (right) states of the capillary rise                           |
+|                                                                                                                   |
++-------------------------------------------------------------------------------------------------------------------+
+
+The quantity of interest of this problem is the difference in height (:math:`\Delta H` in the previous figure) between the tip of the meniscus and the surface of the fluid outside of the central part. The Jurin's law [#liu2018]_ gives an asymptotical value of :math:`\Delta H` :
+
+.. math::
+    \Delta H = \frac{\sigma\cos{\alpha_c}}{\rho_1gR}
+
+with :math:`\sigma` the surface tension coefficient, :math:`\alpha_c` the angle of contact imposed on the wall, :math:`\rho_1` the density of the denser fluid, :math:`g` the gravitational acceleration and :math:`R` the radius of the central part.
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply doc/source/examples/multiphysics/jurins-law/jurins-law.rst]: Good catch, I kept the name radius but half-distance fits better for the problem under study

File: doc/source/examples/multiphysics/jurins-law/jurins-law.rst [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+      set output name       = jurins-law-2d
+      set output frequency  = 10
+      set output path       = ./
+      set max time step     = 5e-4
+      set adapt             = true
+      set max cfl           = 0.8
+      set time end          = 0.5
+      set time step         = 5e-4
+      set output boundaries = true
+    end
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply doc/source/examples/multiphysics/jurins-law/jurins-law.rst:81]: It should be 0.5 everywhere

File: doc/source/examples/multiphysics/jurins-law/jurins-law.rst [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+      end
+    end
+
+-----------------------
+Running the Simulation
+-----------------------
+
+We call lethe-fluid by invoking:
+
+``mpirun -np $number_of_CPU_cores lethe-fluid jurins-law-2d.prm``
```
[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply doc/source/examples/multiphysics/jurins-law/jurins-law.rst]: Very cool, I didn't know about this feature

File: doc/source/examples/multiphysics/jurins-law/images/jurins-law.svg
[the remark this answers is not in the document]
Comment (AmishgaAlphonius) [reply doc/source/examples/multiphysics/jurins-law/images/jurins-law.svg:1]: No, I was asking in the y-direction, if the wall is half-way submerged in both liquid?

[the remark this answers is not in the document]
Comment (PierreLaurentinCS) [reply doc/source/examples/multiphysics/jurins-law/images/jurins-law.svg:1]: Yes it is, you can see that the total height is 8R and the height of the liquid is 4R. So it is halfway submerged

[the remark this answers is not in the document]
Comment (AmishgaAlphonius) [reply doc/source/examples/multiphysics/jurins-law/images/jurins-law.svg:1]: But, there is nothing relating the fluids to the wall's position in y

**Review by AmishgaAlphonius — APPROVED** [review]
Only the comment related to the figure with the y-position of the wall in left :)
Great work!

**Comment by blaisb** [comment]
@PierreLaurentinCS Can you find a second reviewer please?

**Review by oguevremont — CHANGES_REQUESTED** [review]
Great work! only 2-3 typos, and I think the parameter file (and section in the RST) for the BC will need changes

**Comment by blaisb** [comment]
Will merge once I have the ok of @oguevremont
